File: SaintVenant/Ju/main1d.jl
using Printf
using DelimitedFiles




# the type for the domain
struct Domain
    x_start::Float64
    x_end::Float64
end


# Initial solution
function InitSolution!(V::Array{Float64, 2}, domain::Domain)
    nx = size(V, 2)

    hx = (domain.x_end - domain.x_start) / nx
    xc = 0.5  * (domain.x_start + domain.x_end) + nx/8 * hx
    
    for i=1:nx
        x = domain.x_start + (i-0.5) * hx    
        if abs(x - xc) < 0.2
            V[1, i] = 1
        else
            V[1, i] = 0
        end

        V[2, i] = 0
    end
end


# Model
function F!(f, V1, V2, tol)
    f[1] = V2
    if V1 < tol
        f[2] = 0
    else
        f[2] = V2^2 / V1 + 0.5 * 9.81 * V1^2
    end
end


# Scheme
function LaxFriedrich!(V::Array{Float64, 2}, Vold::Array{Float64, 2}, lambdas::Array{Float64, 1}, dt, dx, tol)
    nx = size(V, 2)
    Cx = dt / dx
    
    flux1 = Array{Float64}([0., 0.])
    flux2 = Array{Float64}([0., 0.])


    F!(flux1, Vold[1, 1], Vold[2, 1], tol)

    # left boundary
    V[2, 1] += Cx * (flux1[2] - lambdas[1] * Vold[2, 1])

    # interior edges
    for i=1:nx-1
        F!(flux2, Vold[1, i+1], Vold[2, i+1], tol)
        lambda = max(lambdas[i], lambdas[i+1])
        
        flux = 0.5 * Cx * ((flux2[1] + flux1[1]) - lambda * (Vold[1, i+1] - Vold[1, i]))
        V[1, i] -= flux
        V[1, i+1] += flux
        
        flux = 0.5 * Cx * ((flux2[2] + flux1[2]) - lambda * (Vold[2, i+1] - Vold[2, i]))
        V[2, i] -= flux
        V[2, i+1] += flux
        
        flux1, flux2 = flux2, flux1
    end

    # right boundary
    V[2, nx] -= Cx * (flux1[2] + lambdas[nx] * Vold[2, nx])
end

# Update the eigenvalues lambdas
function update_eigenvalues!(lambdas::Array{Float64, 1}, V::Array{Float64, 2}, tol::Float64)
    nx = size(V, 2)
    
    for i=1:nx
        if V[1, i] < tol
            lambdas[i] = 0.
        else
            lambdas[i] = abs(V[2, i] / V[1, i]) + sqrt(9.81 * V[1, i])
        end
    end
end



function update_to_time(V::Array{Float64, 2}, Vold::Array{Float64, 2}, lambdas::Array{Float64, 1},
                        t::Float64, tframe::Float64, dx::Float64, tol::Float64)
    while(t < tframe)
        @. Vold = V
        update_eigenvalues!(lambdas, Vold, tol)
        dt = min(0.5 * dx / maximum(lambdas), tframe - t)
        LaxFriedrich!(V, Vold, lambdas, dt, dx, tol)
        t += dt
    end
end




# initialization
const domain = Domain(0., 1.)

# loop in time
const T = 2.0
const tol = eps(Float64)

const nstart = 256
const nstep = 10

open("RunningOn" * gethostname(), "w") do file
    for istep=0:nstep-1
        Nx = nstart * 2^istep
        dx = (domain.x_end - domain.x_start) / Nx
    
        V = Array{Float64, 2}(undef, 2, Nx)        
        Vold = Array{Float64, 2}(undef, 2, Nx)
        lambdas = Array{Float64, 1}(undef, Nx)
        
        if istep == 0
            InitSolution!(V, domain)
            print("Warming up... ")
            update_to_time(V, Vold, lambdas, 0., eps(Float64), dx, tol)
            println("Done")
        end
        
        t = 0
        InitSolution!(V, domain)
        
        #println("Initial time t = ", t)
        print("Running with N = ", Nx, " : ")
        t = @elapsed update_to_time(V, Vold, lambdas, 0., T, dx, tol)
        println(t, "s")
        #println("End of simulation, t = ", t)
        
        #println("mean(h) = ", mean(V[1, :]))
        #writedlm("sol-cpu", V[1, :], "\n")
        write(file, string(Nx), "\t", string(t), "\n")
    end
end

File: SaintVenant/Python-Numba/main1d.py
import numpy as np
import numba
import time
import socket


def init_solution(V, domain):
    nx = V.shape[0]
    hx = (domain[1] - domain[0]) / nx
    xc = 0.5 * (domain[0] + domain[1]) + nx / 8 * hx

    V[:, 0] = np.abs(domain[0] + (np.arange(nx) + 0.5) * hx - xc) < 0.2
    V[:, 1] = 0


@numba.njit
def model(ff, V1, V2, tol):
    ff[0] = V2
    if V1 < tol:
        ff[1] = 0.
    else:
        ff[1] = V2**2 / V1 + 0.5 * 9.81 * V1**2


@numba.njit
def LaxFriedrich(V, Vold, lambdas, dt, dx, tol):
    nx = V.shape[0]
    Cx = dt / dx

    flux1 = np.empty(2)
    flux2 = np.empty(2)

    model(flux1, Vold[0, 0], Vold[0, 1], tol)

    V[0, 1] += Cx * (flux1[1] - lambdas[0] * Vold[0, 1])
    for i in range(nx-1):
        model(flux2, Vold[i+1, 0], Vold[i+1, 1], tol)
        lmbd = max(lambdas[i], lambdas[i+1])

        flux = 0.5 * Cx * ((flux2[0] + flux1[0]) - lmbd * (Vold[i+1, 0] - Vold[i, 0]))
        V[i, 0] -= flux
        V[i+1, 0] += flux

        flux = 0.5 * Cx * ((flux2[1] + flux1[1]) - lmbd * (Vold[i+1, 1] - Vold[i, 1]))
        V[i, 1] -= flux
        V[i+1, 1] += flux

        flux1, flux2 = flux2, flux1

    V[-1, 1] -= Cx * (flux1[1] + lambdas[-1] * Vold[-1, 1])
        

@numba.njit
def update_eigenvalues(lambdas, V, tol):
    nx = lambdas.size
    for i in range(nx):
        if V[i, 0] < tol:
            lambdas[i] = 0.
        else:
            lambdas[i] = np.abs(V[i, 1] / V[i, 0]) + np.sqrt(9.81 * V[i, 0])


@numba.njit
def update_to_time(t, tfinal, V, Vold, lambdas, dx, tol):
    while t < tfinal:
        for i in range(V.shape[0]):
            Vold[i, 0] = V[i, 0]
            Vold[i, 1] = V[i, 1]
            
        update_eigenvalues(lambdas, Vold, tol)
        dt = min(0.5 * dx / lambdas.max(), tfinal - t)

        LaxFriedrich(V, Vold, lambdas, dt, dx, tol)

        t += dt
        
    return t



domain = np.array([0, 1])
T = 2.0
nstart = 256
nstep = 10

with open("RunningOn" + socket.gethostname(), "w") as f:
    for i in range(nstep):
        Nx = nstart * 2**i
        dx = (domain[1] - domain[0]) / Nx

        t = 0.
        V = np.empty((Nx, 2))
        Vold = np.empty_like(V)
        lambdas = np.empty(Nx)

        tol = np.finfo(np.float64).eps

        if i == 0:
            init_solution(V, domain)
            print("Warming up...", end=" ")
            update_to_time(t, tol, V, Vold, lambdas, dx, tol)
            print("Done")

        init_solution(V, domain)
        
        #print(f"Initial time t = {t}")
        print(f"Running with N = {Nx} :", end=" ")
        t_start = time.time()
        t = update_to_time(t, T, V, Vold, lambdas, dx, tol)
        t_end = time.time()
        print(f"{t_end - t_start}")
        #print(f"End of simulation, t = {t}")
        #print(f"Elapsed time: {t_end - t_start}")

        #print(f"mean(h) = {np.mean(V[:, 0])}")
        #np.savetxt("sol-cpu", V[:, 0])
        f.write(str(Nx) + "\t" + str(t_end - t_start) + "\n")
